### ISSUE TITLE
[Bug]:  Federation not working (with one exception)

### ISSUE THREAD

**Issue by life8853** [LemmyNet/lemmy#3167]
### Requirements

- [X] Is this a bug report? For questions or discussions use https://lemmy.ml/c/lemmy_support
- [X] Did you check to see if this issue already exists?
- [X] Is this only a single bug? Do not put multiple bugs in one issue.
- [ ] Is this a UI / front end issue? Use the [lemmy-ui](https://github.com/LemmyNet/lemmy-ui) repo.

### Summary

Today i set up a brand new instance and federation (according to https://join-lemmy.org/docs/en/administration/troubleshooting.html) should work. Other instances are able to "see" my instance but not the other way around.

Update the situation is more interesting, only some instances are able to see mine:
- https://lemmyunchained.net/c/main@tslemmy.duckdns.org Works
- https://lemmy.world/c/main@tslemmy.duckdns.org 404
- https://lemmy.ml/c/main@tslemmy.duckdns.org 404
- https://feddit.nl/c/main@tslemmy.duckdns.org 404

Otherway:
- https://tslemmy.duckdns.org/c/guitarporn@lemmyunchained.net 404

### Steps to Reproduce

1. Set up an instance i guess?
2.
3.

### Technical Details

Unblocked ports on the VPS: 80 and 443

Error that i get
```
lemmy-lemmy-1     | 2023-06-17T12:38:01.989561Z ERROR HTTP request{http.method=GET http.scheme="http" http.host=tslemmy.duckdns.org http.target=/api/v3/ws otel.kind="server" request_id=6378fd3e-890c-4db3-94e3-02ab0569468a http.status_code=101 otel.status_code="OK"}: lemmy_server::api_routes_websocket: couldnt_find_object: Request error: error sending request for url (https://lemmy.world/c/selfhosted): operation timed out
lemmy-lemmy-1     |    0: lemmy_apub::fetcher::search::search_query_to_object_id
lemmy-lemmy-1     |              at crates/apub/src/fetcher/search.rs:17
lemmy-lemmy-1     |    1: lemmy_apub::api::resolve_object::perform
lemmy-lemmy-1     |            with self=ResolveObject { q: "https://lemmy.world/c/selfhosted", auth: Some(Sensitive) }
lemmy-lemmy-1     |              at crates/apub/src/api/resolve_object.rs:21
lemmy-lemmy-1     |    2: lemmy_server::root_span_builder::HTTP request
lemmy-lemmy-1     |            with http.method=GET http.scheme="http" http.host=tslemmy.duckdns.org http.target=/api/v3/ws otel.kind="server" request_id=6378fd3e-890c-4db3-94e3-02ab0569468a http.status_code=101 otel.status_code="OK"
lemmy-lemmy-1     |              at src/root_span_builder.rs:16
```

docker-compose.yml

```
version: "3.3"

networks:
  # communication to web and clients
  lemmyexternalproxy:
  # communication between lemmy services
  lemmyinternal:
    driver: bridge
    internal: true

services:
  proxy:
    image: nginx:1-alpine
    networks:
      - lemmyinternal
      - lemmyexternalproxy
    ports:
      # only ports facing any connection from outside
      - 80:80
      - 443:443
    volumes:
      - ./nginx.conf:/etc/nginx/nginx.conf:ro
      # setup your certbot and letsencrypt config
      - ./certbot:/var/www/certbot
      - /etc/letsencrypt/:/etc/letsencrypt/:ro
    restart: always
    depends_on:
      - pictrs
      - lemmy-ui

  lemmy:
    image: dessalines/lemmy:0.17.4
    hostname: lemmy
    networks:
      - lemmyinternal
    restart: always
    environment:
      - RUST_LOG="warn,lemmy_server=info,lemmy_api=info,lemmy_api_common=info,lemmy_api_crud=info,lemmy_apub=info,lemmy_db_schema=info,lemmy_db_views=info,lemmy_db_views_actor=info,lemmy_db_views_moderator=info,lemmy_routes=info,lemmy_utils=info,lemmy_websocket=info"
    volumes:
      - ./lemmy.hjson:/config/config.hjson
    depends_on:
      - postgres
      - pictrs

  lemmy-ui:
    image: dessalines/lemmy-ui:0.17.4
    networks:
      - lemmyinternal
    environment:
      # this needs to match the hostname defined in the lemmy service
      - LEMMY_UI_LEMMY_INTERNAL_HOST=lemmy:8536
      # set the outside hostname here
      - LEMMY_UI_LEMMY_EXTERNAL_HOST=lemmy-ui:1236
      - LEMMY_HTTPS=true
    depends_on:
      - lemmy
    restart: always

  pictrs:
    image: asonix/pictrs:0.3.1
    # this needs to match the pictrs url in lemmy.hjson
    hostname: pictrs
    # we can set options to pictrs like this, here we set max. image size and forced format for conversion
    # entrypoint: /sbin/tini -- /usr/local/bin/pict-rs -p /mnt -m 4 --image-format webp
    networks:
      - lemmyinternal
    environment:
      - PICTRS__API_KEY=api-key
    user: 991:991
    volumes:
      - ./volumes/pictrs:/mnt
    restart: always

  postgres:
    image: postgres:15-alpine
    # this needs to match the database host in lemmy.hson
    hostname: postgres
    networks:
      - lemmyinternal
    environment:
      - POSTGRES_USER=lemmy
      - POSTGRES_PASSWORD=passwd
      - POSTGRES_DB=lemmy
    volumes:
      - ./volumes/postgres:/var/lib/postgresql/data
    restart: always
```

nginx.conf 

```
worker_processes 1;
events {
    worker_connections 1024;
}
http {
    limit_req_zone $binary_remote_addr zone=tslemmy.duckdns.org_ratelimit:10m rate=1r/s;
    upstream lemmy {
        # this needs to map to the lemmy (server) docker service hostname
        server "lemmy:8536";
    }
    upstream lemmy-ui {
        # this needs to map to the lemmy-ui docker service hostname
        server "lemmy-ui:1234";
    }
    server {
    # allow letsencrypt challenge
    # redirect everything else to 443
    listen 80;
    listen [::]:80;
    server_name tslemmy.duckdns.org;
    location /.well-known/acme-challenge/ {
        root /var/www/certbot;
    }
    location / {
        return 301 https://$host$request_uri;
    }
    }

    server {
    listen 443 ssl;
    listen [::]:443 ssl;
    http2 on;
    server_name tslemmy.duckdns.org;

    ssl_certificate /etc/letsencrypt/live/tslemmy.duckdns.org/fullchain.pem;
    ssl_certificate_key /etc/letsencrypt/live/tslemmy.duckdns.org/privkey.pem;

    # Various TLS hardening settings
    # https://raymii.org/s/tutorials/Strong_SSL_Security_On_nginx.html
    ssl_protocols TLSv1.2 TLSv1.3;
    ssl_prefer_server_ciphers on;
    ssl_ciphers 'ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256';
    ssl_session_timeout  10m;
    ssl_session_cache shared:SSL:10m;
    ssl_session_tickets on;
    ssl_stapling on;
    ssl_stapling_verify on;

    # Hide nginx version
    server_tokens off;

    # Enable compression for JS/CSS/HTML bundle, for improved client load times.
    # It might be nice to compress JSON, but leaving that out to protect against potential
    # compression+encryption information leak attacks like BREACH.
    gzip on;
    gzip_types text/css application/javascript image/svg+xml;
    gzip_vary on;

    # Only connect to this site via HTTPS for the two years
    add_header Strict-Transport-Security "max-age=63072000";

    # Various content security headers
    add_header Referrer-Policy "same-origin";
    add_header X-Content-Type-Options "nosniff";
    add_header X-Frame-Options "DENY";
    add_header X-XSS-Protection "1; mode=block";

    # Upload limit for pictrs
    client_max_body_size 20M;

    # frontend
    location / {
        # distinguish between ui requests and backend
        # don't change lemmy-ui or lemmy here, they refer to the upstream definitions on top
        set $proxpass "http://lemmy-ui";

        if ($http_accept = "application/activity+json") {
            set $proxpass "http://lemmy";
        }
        if ($http_accept = "application/ld+json; profile=\"https://www.w3.org/ns/activitystreams\"") {
            set $proxpass "http://lemmy";
        }
        if ($request_method = POST) {
            set $proxpass "http://lemmy";
        }
        proxy_pass $proxpass;

        rewrite ^(.+)/+$ $1 permanent;

        # Send actual client IP upstream
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header Host $host;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    }

    # backend
    location ~ ^/(api|feeds|nodeinfo|.well-known) {
        proxy_pass "http://lemmy";
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "upgrade";

        # Rate limit
        limit_req zone=tslemmy.duckdns.org_ratelimit burst=30 nodelay;

        # Add IP forwarding headers
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header Host $host;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    }

    # pictrs only - for adding browser cache control.
    location ~ ^/(pictrs) {
        # allow browser cache, images never update, we can apply long term cache
        expires 120d;
        add_header Pragma "public";
        add_header Cache-Control "public";

        proxy_pass "http://lemmy";
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "upgrade";

        # Rate limit
        limit_req zone=tslemmy.duckdns.org_ratelimit burst=30 nodelay;

        # Add IP forwarding headers
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header Host $host;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    }

    # Redirect pictshare images to pictrs
    location ~ /pictshare/(.*)$ {
        return 301 /pictrs/image/$1;
    }
}
}
```

lemmy.hjson

```
{
  # for more info about the config, check out the documentation
  # https://join-lemmy.org/docs/en/administration/configuration.html
  # only few config options are covered in this example config

  setup: {
    # username for the admin user
    admin_username: "username"
    # password for the admin user
    admin_password: "admin_passwd"
    # name of the site (can be changed later)
    site_name: "TSowelemmy"
  }

  # the domain name of your instance (eg "lemmy.ml")
  hostname: "tslemmy.duckdns.org"
  # address where lemmy should listen for incoming requests
  bind: "0.0.0.0"
  # port where lemmy should listen for incoming requests
  port: 8536
  # Whether the site is available over TLS. Needs to be true for federation to work.
  tls_enabled: true

  # pictrs host
  pictrs: {
    url: "http://pictrs:8080/"
    api_key: "apikey"
  }

  # settings related to the postgresql database
  database: {
    # name of the postgres database for lemmy
    database: "lemmy"
    # username to connect to postgres
    user: "lemmy"
    # password to connect to postgres
    password: "passwd"
    # host where postgres is running
    host: "postgres"
    # port where postgres can be accessed
    port: 5432
    # maximum number of active sql connections
    pool_size: 5
  }
}
```

### Version

BE 0.17.4

### Lemmy Instance URL

tslemmy.duckdns.org

**Comment by RocketDerp** [comment]
> Today i set up a brand new instance

A user of your instance has to 'prime' each community by doing a 'join/subscribe' on community to start federation data flowing from the remote instance to your instance.

It will backfill about 10 postings, but does not seem to backfill comments.

I'm not sure how user profiles get brought over between instances, I suspect that may be on demand?

**Comment by life8853** [comment]
> A user of your instance has to 'prime' each community by doing a 'join/subscribe' on community to start federation data flowing from the remote instance to your instance.

I know but i can't find any other instances from mine

**Comment by RocketDerp** [comment]
> I know but i can't find any other instances from mine

Do you know how to find communities of remote instances from your own community? it isn't an  intuitive process IMHO.

You have to use the "all content search" feature in the upper-right of the lemmy-ui webapp and search for the rull url.

I have a community for instance admins to talk about Lemmy technical issues with performance, try to find that by searching your local site with this url in the search box: https://lemmy.ml/c/lemmyperformance

A remote profile might be a good way, because I think that is more automatic/on-demand. For example, this is one instance showing another instance profile: https://lemmy.world/u/RoundSparrow@lemmy.ml

**Comment by RocketDerp** [comment]
P.S. do you know Google Chrome browser flags your domain as 'deceptive'?

**Comment by RocketDerp** [comment]
I went to your instance directly, I suggest you put at least one posting and comment in your 'main' community so we can see if data is flowing.

I was able to discover your 'main' community from a remote instance, but when I join/subscribe it is 'pending' - which is a sign data isn't flowing bidirectional.

**Comment by life8853** [comment]
> I went to your instance directly, I suggest you put at least one posting and comment in your 'main' community so we can see if data is flowing.
> 
> I was able to discover your 'main' community from a remote instance, but when I join/subscribe it is 'pending' - which is a sign data isn't flowing bidirectional.

i did do a post and a comment

**Comment by RocketDerp** [comment]
Yha, from your server I can't seem to get outbound working... I tried to discover a remote community:  https://tslemmy.duckdns.org/search/q/https%3A%2F%2Flemmy.world%2Fc%2Fguitars/type/All/sort/TopAll/listing_type/All/community_id/0/creator_id/0/page/1

And it doesn't pop up.  You are using Docker, I wonder if your docker isn't able to do outbound from lemmy_server app? Maybe your "lemmyexternalproxy" value? I'm not using Docker, so I don't have experience with it.

**Comment by penttinj** [comment]
I'm having the same issue. I can subscribe to my own community from fe. lemmy.world, but I can't retrieve remote communities from my own instance.
The documentation is a mess, the join-lemmy.org instructions don't work or are wrong. For example, the reverse proxy instructions say to add `proxy_set_header` but that crashes my nginx server because `include proxy_params;` already includes `proxy_set_header`.

Anyways sorry if it came off negative, this is a huge project by few contributors which others get to enjoy for free.

I copied this from the lemmy(backend?) docker service logs after making a community search:
```
2023-06-17T17:26:49.903782Z ERROR HTTP request{http.method=GET http.scheme="https" http.host=lemmy http.target=/api/v3/ws otel.kind="server" request_id=30ec3856-8b6e-4a4d-87db-9b54ee3e68c5 http.status_code=101 otel.status_code="OK"}: lemmy_server::api_routes_websocket: couldnt_find_object: Request error: error sending request for url (https://beehaw.org/c/programming): operation timed out
   0: lemmy_apub::fetcher::search::search_query_to_object_id
             at crates/apub/src/fetcher/search.rs:17
   1: lemmy_apub::api::resolve_object::perform
           with self=ResolveObject { q: "https://beehaw.org/c/programming", auth: Some(Sensitive) }
             at crates/apub/src/api/resolve_object.rs:21
   2: lemmy_server::root_span_builder::HTTP request
           with http.method=GET http.scheme="https" http.host=lemmy http.target=/api/v3/ws otel.kind="server" request_id=30ec3856-8b6e-4a4d-87db-9b54ee3e68c5 http.status_code=101 otel.status_code="OK"
             at src/root_span_builder.rs:16
```

And a second request for the same community which resulted in a warning instead of an error:
```
2023-06-17T17:27:14.401714Z  WARN Error encountered while processing the incoming HTTP request: lemmy_server::root_span_builder: couldnt_find_object: Request error: error sending request for url (https://beehaw.org/c/programming): operation timed out
   0: lemmy_apub::fetcher::search::search_query_to_object_id
             at crates/apub/src/fetcher/search.rs:17
   1: lemmy_apub::api::resolve_object::perform
           with self=ResolveObject { q: "https://beehaw.org/c/programming", auth: Some(Sensitive) }
             at crates/apub/src/api/resolve_object.rs:21
   2: lemmy_server::root_span_builder::HTTP request
           with http.method=GET http.scheme="http" http.host=lemmy-ui http.target=/api/v3/resolve_object otel.kind="server" request_id=387a49e8-35b3-4020-ae86-9574e1f83e43 http.status_code=400 otel.status_code="OK"
             at src/root_span_builder.rs:16
LemmyError { message: Some("couldnt_find_object"), inner: Request error: error sending request for url (https://beehaw.org/c/programming): operation timed out

Caused by:
    0: error sending request for url (https://beehaw.org/c/programming): operation timed out
    1: operation timed out, context: "SpanTrace" }
```

Sure seems like outbound requests are blocked. I'm running nginx which forwards to docker, then docker forwards that to the internal nginx proxy. I'm using the nginx.conf from this repo's main branch. SSL is handled by my own nginx reverse proxy.

**Comment by penttinj** [comment]
Whelp. I removed the `internal` property from the docker-compose yaml in the networks section and started finding remote communities.

`docker-compose.yml`
```
networks:
  # communication to web and clients
  lemmyexternalproxy:
  # communication between lemmy services
  lemmyinternal:
    driver: bridge
```

I don't know if there's any security concerns with this. Do it on your own risk.

**Comment by life8853** [comment]
@penttinj I solved it too and my solutions seems more secure. I think you should do it this way

I added the external proxy network to "lemmy" container 

```
lemmy:
    image: dessalines/lemmy:0.17.4
    hostname: lemmy
    networks:
      - lemmyinternal
      - lemmyexternalproxy
    restart: always
```

**Comment by RocketDerp** [comment]
See also, closed Issue #2990

**Comment by life8853** [comment]
Basically what i did
